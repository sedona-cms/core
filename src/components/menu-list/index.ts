import MenuList from './menu-list'

export { MenuList }
