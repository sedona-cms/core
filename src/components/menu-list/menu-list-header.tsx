import Vue, { VNode, PropType, CreateElement } from 'vue'

export default Vue.extend({
  name: 'MenuListHeader',
  functional: true,
  props: {
    title: {
      type: String as PropType<string>,
      required: true,
    },
  },
  render(_h: CreateElement, { props }): VNode {
    return <q-item-label header={true}>{props.title}</q-item-label>
  },
})
