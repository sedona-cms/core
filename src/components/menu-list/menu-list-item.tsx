import Vue, { VNode, PropType } from 'vue'

export default Vue.extend({
    name: 'MenuListItem',
    props: {
        id: {
            type: String as PropType<string>,
            required: true,
        },
        title: {
            type: String as PropType<string>,
            default: '<NO TITLE>',
        },
        subTitle: {
            type: String as PropType<string>,
            default: '',
        },
        icon: {
            type: String as PropType<string>,
            default: 'folder',
        },
        component: {
            type: String as PropType<string>,
            default: '',
        },
        items: {
            type: Array as PropType<MenuItem[]>,
            default: () => [],
        },
        type: {
            type: String as PropType<string>,
            default: 'item',
            validator: (value) => [ 'item', 'section' ].includes(value),
        },
    },
    methods: {
        menuItemClick(): void {
            const data: any = {
                id: this.id,
                title: this.title,
                subTitle: this.subTitle,
                icon: this.icon,
                type: this.type,
                component: undefined,
            }
            if (this.type === 'item') {
                data.component = this.component
            } else {
                data.items = this.items
            }
            this.$root.$emit('admin:view-change', data)
        },
    },
    render(): VNode {
        let subTitle: string | undefined
        if (this.subTitle !== '') {
            subTitle = <q-item-label caption={ true }>{ this.subTitle }</q-item-label>
        }

        const iconArrow: VNode | undefined =
            this.type === 'section' ? (
                <q-item-section side={ true }>
                    <q-icon name="keyboard_arrow_right"/>
                </q-item-section>
            ) : undefined

        return (
            <q-item clickable={ true } on-click={ this.menuItemClick }>
                <q-item-section avatar={ true }>
                    <q-avatar icon={ this.icon } color="grey-7" text-color="white"/>
                </q-item-section>
                <q-item-section>
                    <q-item-label lines={ 1 }>{ this.title }</q-item-label>
                    { subTitle }
                </q-item-section>
                { iconArrow }
            </q-item>
        )
    },
})
