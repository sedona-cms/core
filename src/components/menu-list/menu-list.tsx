import Vue, { PropType, VNode, CreateElement } from 'vue'
import MenuListItem from './menu-list-item'
import MenuListHeader from './menu-list-header'
import MenuListSection from './menu-list-section'

function isConditionTrue(menuItem: MenuItem): boolean {
  if (menuItem.conditions === undefined) return true
  for (const item of menuItem.conditions) {
    if (item.type === '=') {
      if (item.field === 'meta') {
        // const result = this.$route[item.field] === item.value
      } else {
        // this.$route[item.field] === item.value
      }
    }
  }
  return false
}

export default Vue.extend({
  name: 'MenuList',
  components: {
    MenuListItem,
    MenuListHeader,
    MenuListSection,
  },
  props: {
    items: {
      type: Array as PropType<MenuItem[]>,
      required: true,
    },
  },
  computed: {
    menuItems(): MenuItem[] {
      const items: MenuItem[] = []
      for (const menuItem of this.items) {
        switch (menuItem.type) {
          case 'item': {
            if (menuItem.conditions === undefined || isConditionTrue.call(this, menuItem)) {
              items.push(menuItem)
            }
            break
          }
          case 'section':
            break
        }
      }
      return items
    },
  },
  render(h: CreateElement): VNode {
    const childComponents = new Set<VNode>()

    function createMenuItem(item: MenuItem): VNode {
      const props: MenuItem = {
        id: item.id,
        ...(item.title !== undefined ? { title: item.title } : { title: '<NO TITLE>' }),
        ...(item.type !== undefined ? { type: item.type } : { type: 'item' }),
      }

      switch (props.type) {
        case 'item': {
          props.subTitle = (item as SimpleMenuItem).subTitle ?? ''
          props.icon = (item as SimpleMenuItem).icon ?? 'folder'
          props.component = (item as SimpleMenuItem).component
          props.params = (item as SimpleMenuItem).params ?? {}
          break
        }
        case 'section': {
          props.subTitle = (item as SectionMenuItem).subTitle ?? ''
          props.icon = (item as SectionMenuItem).icon ?? 'folder'
          props.items = (item as SectionMenuItem).items ?? []
          break
        }
      }

      return h('menu-list-item', { props })
    }

    function createMenuSection(item: SectionMenuItem): VNode[] {
      const result: VNode[] = [<q-item-label header={true}>{item.title}</q-item-label>]
      if (!Array.isArray(item.items)) {
        return result
      }
      for (const menuItem of item.items) {
        result.push(createMenuItem(menuItem))
      }

      return result
    }

    this.items.forEach((item: MenuItem) => {
      switch (item.type) {
        case 'item':
          childComponents.add(createMenuItem(item))
          break
        case 'header':
          childComponents.add(<menu-list-header title={item.title} />)
          break
        case 'section':
          childComponents.add(
            <menu-list-section
              id={item.id}
              title={item.title}
              sub-title={item.subTitle ?? ''}
              icon={item.icon ?? 'folder'}
              items={item.items ?? []}
            />
          )
          break
      }
    })

    return (
      <q-list padding={true} dark={true} class="full-width">
        {[...childComponents]}
      </q-list>
    )
  },
})
