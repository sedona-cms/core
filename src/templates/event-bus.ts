import Vue from 'vue'

const vueInstance = new Vue()
const eventBus = {
  emit(event: string, ...args: any[]): void {
    vueInstance.$emit(event, args)
  },
  on(event: string | string[], callback: Function): void {
    vueInstance.$on(event, callback)
  },
  off(event?: string | string[], callback?: Function): void {
    vueInstance.$off(event, callback)
  },
  once(event: string | string[], callback: Function): void {
    vueInstance.$once(event, callback)
  },
}

export { eventBus }
