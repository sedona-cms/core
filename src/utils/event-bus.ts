import Vue from 'vue'

type EventCallback = (...args: never[]) => void

const vueInstance = new Vue()
const eventBus = {
  emit(event: string, ...args: unknown[]): void {
    vueInstance.$emit(event, args)
  },
  on(event: string | string[], callback: EventCallback): void {
    vueInstance.$on(event, callback)
  },
  off(event?: string | string[], callback?: EventCallback): void {
    vueInstance.$off(event, callback)
  },
  once(event: string | string[], callback: EventCallback): void {
    vueInstance.$once(event, callback)
  },
}

export { eventBus }
