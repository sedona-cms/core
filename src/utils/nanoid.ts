import { customAlphabet } from 'nanoid'

const alphabet = '0123456789abcdefghijklmnopqrstuvwxyz'
const nanoid = customAlphabet(alphabet, 10)

/**
 * Generate unique identifier
 *
 * @exports
 * @returns {string} short unique id
 */
export function generateId(): string {
  return nanoid()
}
