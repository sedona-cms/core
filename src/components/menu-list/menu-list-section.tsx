import { VNode, PropType } from 'vue'
import mixins from 'vue-typed-mixins'
import { menuListItemMixin } from './mixins/menu-list-item-mixin'
import { eventBus } from '../../utils/event-bus'

export default mixins(menuListItemMixin).extend({
  name: 'MenuListSection',
  props: {
    items: {
      type: Array as PropType<MenuItem[]>,
      required: true,
    },
  },
  methods: {
    menuItemClick(): void {
      const item: SectionMenuItem = {
        type: 'section',
        id: this.id,
        title: this.title,
        subTitle: this.subTitle,
        icon: this.icon,
        items: this.items,
      }
      eventBus.emit('core:navigate', item)
    },
  },
  render(): VNode {
    let subTitle: string | undefined
    if (this.subTitle !== '') {
      subTitle = <q-item-label caption={true}>{this.subTitle}</q-item-label>
    }

    return (
      <q-item clickable={true} on-click={this.menuItemClick}>
        <q-item-section avatar={true}>
          <q-avatar icon={this.icon} color="grey-7" text-color="white" />
        </q-item-section>
        <q-item-section>
          <q-item-label lines={1}>{this.title}</q-item-label>
          {subTitle}
        </q-item-section>
        <q-item-section side={true}>
          <q-icon name="keyboard_arrow_right" />
        </q-item-section>
      </q-item>
    )
  },
})
