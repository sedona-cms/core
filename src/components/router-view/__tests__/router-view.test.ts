import { mount, createLocalVue } from '@vue/test-utils'
import {
  QAvatar,
  QIcon,
  QItem,
  QItemLabel,
  QItemSection,
  QList,
  QScrollArea,
  QTabPanels,
  QTabPanel,
  QSeparator,
  QToolbarTitle,
  QBtn,
  QToolbar,
  Quasar,
} from 'quasar'
import { eventBus } from '../../../utils/event-bus'
import { generateId } from '../../../utils/nanoid'
import { MenuList } from '../../menu-list'
import { RouterView } from '..'

const localVue = createLocalVue()

localVue.use(Quasar, {
  components: {
    QItemLabel,
    QItemSection,
    QIcon,
    QItem,
    QAvatar,
    QList,
    QScrollArea,
    QTabPanels,
    QTabPanel,
    QSeparator,
    QToolbarTitle,
    QBtn,
    QToolbar,
  },
})

describe('RouterView', () => {
  const menuItems = [
    {
      id: generateId(),
      title: 'Posts',
      type: 'item',
      component: 'posts-view',
    },
    {
      id: generateId(),
      title: 'Meta Section',
      type: 'section',
      items: [
        {
          id: generateId(),
          title: 'Submenu Posts',
          type: 'item',
          component: 'posts-view',
        },
      ],
    },
  ]
  const wrapper = mount(RouterView, {
    localVue,
    mocks: {
      $sedona: {
        menuItems,
      },
    },
  })

  test('it can be created with home menu list', () => {
    expect(wrapper.exists()).toBeTruthy()

    const menu = wrapper.findComponent(MenuList)

    expect(menu.exists()).toBeTruthy()
    expect(menu.vm.menuItems).toHaveLength(2)
    expect(menu.vm.menuItems[0]).toEqual(menuItems[0])
    expect(menu.vm.menuItems[1]).toEqual(menuItems[1])
  })

  test('navigate to component', async () => {
    eventBus.emit('core:navigate', menuItems[0])
    await new Promise(resolve => setTimeout(resolve, 500))

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('navigate to section', async () => {
    eventBus.emit('core:navigate', menuItems[1])
    await new Promise(resolve => setTimeout(resolve, 500))

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('to click on menu item', async () => {
    wrapper.find('.q-item').trigger('click')
    await new Promise(resolve => setTimeout(resolve, 500))

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('it can return to home after navigation', async () => {
    eventBus.emit('core:navigate', menuItems[1])
    await new Promise(resolve => setTimeout(resolve, 500))

    eventBus.emit('core:navigate', 'tab-home')
    await new Promise(resolve => setTimeout(resolve, 1000))

    const menu = wrapper.findComponent(MenuList)

    expect(menu.exists()).toBeTruthy()
    expect(menu.vm.menuItems).toEqual(menuItems)
  })
})
