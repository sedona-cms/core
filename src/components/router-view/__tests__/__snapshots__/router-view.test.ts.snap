// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouterView navigate to component 1`] = `
"<div class=\\"q-scrollarea q-scrollarea--dark\\" style=\\"height: calc(100% - 50px); width: 100%; max-width: 300px;\\">
  <div class=\\"scroll relative-position fit hide-scrollbar\\">
    <div class=\\"absolute full-width\\"><object tabindex=\\"-1\\" type=\\"text/html\\" data=\\"about:blank\\" aria-hidden=\\"true\\" style=\\"display: block; position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; height: 100%; width: 100%; overflow: hidden; pointer-events: none; z-index: -1;\\"></object>
      <div class=\\"q-tab-panels q-panel-parent\\">
        <transition-stub name=\\"q-transition--slide-left\\">
          <div role=\\"tabpanel\\" class=\\"q-panel scroll\\">
            <div class=\\"q-tab-panel\\" style=\\"padding: 5px 0px;\\">
              <div class=\\"fit\\">
                <div class=\\"q-toolbar row no-wrap items-center\\"><button tabindex=\\"0\\" type=\\"button\\" role=\\"button\\" class=\\"q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round q-btn--actionable q-focusable q-hoverable q-btn--wrap q-btn--dense\\"><span class=\\"q-focus-helper\\"></span><span class=\\"q-btn__wrapper col row q-anchor--skip\\"><span class=\\"q-btn__content text-center col items-center q-anchor--skip justify-center row\\"><i aria-hidden=\\"true\\" role=\\"img\\" class=\\"material-icons q-icon notranslate\\">folder</i></span></span></button>
                  <div class=\\"q-toolbar__title ellipsis col-shrink\\">Posts</div>
                </div>
                <hr role=\\"separator\\" aria-orientation=\\"horizontal\\" class=\\"q-separator q-separator q-separator--horizontal q-separator--dark\\" horizontal=\\"true\\">
                <div>
                  Posts View
                </div>
              </div>
            </div>
          </div>
        </transition-stub>
      </div>
    </div>
    <!---->
  </div><object tabindex=\\"-1\\" type=\\"text/html\\" data=\\"about:blank\\" aria-hidden=\\"true\\" style=\\"display: block; position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; height: 100%; width: 100%; overflow: hidden; pointer-events: none; z-index: -1;\\"></object>
  <div aria-hidden=\\"true\\" class=\\"q-scrollarea__bar q-scrollarea__bar--v absolute-right q-scrollarea__bar--invisible\\"></div>
  <div aria-hidden=\\"true\\" class=\\"q-scrollarea__thumb q-scrollarea__thumb--v absolute-right q-scrollarea__thumb--invisible\\" style=\\"height: 50px;\\"></div>
</div>"
`;

exports[`RouterView navigate to section 1`] = `
"<div class=\\"q-scrollarea q-scrollarea--dark\\" style=\\"height: calc(100% - 50px); width: 100%; max-width: 300px;\\">
  <div class=\\"scroll relative-position fit hide-scrollbar\\">
    <div class=\\"absolute full-width\\"><object tabindex=\\"-1\\" type=\\"text/html\\" data=\\"about:blank\\" aria-hidden=\\"true\\" style=\\"display: block; position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; height: 100%; width: 100%; overflow: hidden; pointer-events: none; z-index: -1;\\"></object>
      <div class=\\"q-tab-panels q-panel-parent\\">
        <transition-stub name=\\"q-transition--slide-left\\">
          <div role=\\"tabpanel\\" class=\\"q-panel scroll\\">
            <div class=\\"q-tab-panel\\" style=\\"padding: 5px 0px;\\">
              <div class=\\"fit\\">
                <div class=\\"q-toolbar row no-wrap items-center\\"><button tabindex=\\"0\\" type=\\"button\\" role=\\"button\\" class=\\"q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round q-btn--actionable q-focusable q-hoverable q-btn--wrap q-btn--dense\\"><span class=\\"q-focus-helper\\"></span><span class=\\"q-btn__wrapper col row q-anchor--skip\\"><span class=\\"q-btn__content text-center col items-center q-anchor--skip justify-center row\\"><i aria-hidden=\\"true\\" role=\\"img\\" class=\\"material-icons q-icon notranslate\\">folder</i></span></span></button>
                  <div class=\\"q-toolbar__title ellipsis col-shrink\\">Meta Section</div>
                </div>
                <hr role=\\"separator\\" aria-orientation=\\"horizontal\\" class=\\"q-separator q-separator q-separator--horizontal q-separator--dark\\" horizontal=\\"true\\">
                <div class=\\"q-list q-list--dark q-list--padding full-width\\">
                  <div tabindex=\\"0\\" class=\\"q-item q-item-type row no-wrap q-item--clickable q-link cursor-pointer q-focusable q-hoverable\\">
                    <div tabindex=\\"-1\\" class=\\"q-focus-helper\\"></div>
                    <div class=\\"q-item__section column q-item__section--avatar q-item__section--side justify-center\\">
                      <div class=\\"q-avatar\\">
                        <div class=\\"q-avatar__content row flex-center overflow-hidden bg-grey-7 text-white q-chip--colored\\"><i aria-hidden=\\"true\\" role=\\"presentation\\" class=\\"material-icons q-icon notranslate\\">folder</i></div>
                      </div>
                    </div>
                    <div class=\\"q-item__section column q-item__section--main justify-center\\">
                      <div class=\\"q-item__label ellipsis\\">Submenu Posts</div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </transition-stub>
      </div>
    </div>
    <!---->
  </div><object tabindex=\\"-1\\" type=\\"text/html\\" data=\\"about:blank\\" aria-hidden=\\"true\\" style=\\"display: block; position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; height: 100%; width: 100%; overflow: hidden; pointer-events: none; z-index: -1;\\"></object>
  <div aria-hidden=\\"true\\" class=\\"q-scrollarea__bar q-scrollarea__bar--v absolute-right q-scrollarea__bar--invisible\\"></div>
  <div aria-hidden=\\"true\\" class=\\"q-scrollarea__thumb q-scrollarea__thumb--v absolute-right q-scrollarea__thumb--invisible\\" style=\\"height: 50px;\\"></div>
</div>"
`;

exports[`RouterView to click on menu item 1`] = `
"<div class=\\"q-scrollarea q-scrollarea--dark\\" style=\\"height: calc(100% - 50px); width: 100%; max-width: 300px;\\">
  <div class=\\"scroll relative-position fit hide-scrollbar\\">
    <div class=\\"absolute full-width\\"><object tabindex=\\"-1\\" type=\\"text/html\\" data=\\"about:blank\\" aria-hidden=\\"true\\" style=\\"display: block; position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; height: 100%; width: 100%; overflow: hidden; pointer-events: none; z-index: -1;\\"></object>
      <div class=\\"q-tab-panels q-panel-parent\\">
        <transition-stub name=\\"q-transition--slide-left\\">
          <div role=\\"tabpanel\\" class=\\"q-panel scroll\\">
            <div class=\\"q-tab-panel\\" style=\\"padding: 5px 0px;\\">
              <div class=\\"fit\\">
                <div class=\\"q-toolbar row no-wrap items-center\\"><button tabindex=\\"0\\" type=\\"button\\" role=\\"button\\" class=\\"q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round q-btn--actionable q-focusable q-hoverable q-btn--wrap q-btn--dense\\"><span class=\\"q-focus-helper\\"></span><span class=\\"q-btn__wrapper col row q-anchor--skip\\"><span class=\\"q-btn__content text-center col items-center q-anchor--skip justify-center row\\"><i aria-hidden=\\"true\\" role=\\"img\\" class=\\"material-icons q-icon notranslate\\">folder</i></span></span></button>
                  <div class=\\"q-toolbar__title ellipsis col-shrink\\">Submenu Posts</div>
                </div>
                <hr role=\\"separator\\" aria-orientation=\\"horizontal\\" class=\\"q-separator q-separator q-separator--horizontal q-separator--dark\\" horizontal=\\"true\\">
                <div>
                  Posts View
                </div>
              </div>
            </div>
          </div>
        </transition-stub>
      </div>
    </div>
    <!---->
  </div><object tabindex=\\"-1\\" type=\\"text/html\\" data=\\"about:blank\\" aria-hidden=\\"true\\" style=\\"display: block; position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; height: 100%; width: 100%; overflow: hidden; pointer-events: none; z-index: -1;\\"></object>
  <div aria-hidden=\\"true\\" class=\\"q-scrollarea__bar q-scrollarea__bar--v absolute-right q-scrollarea__bar--invisible\\"></div>
  <div aria-hidden=\\"true\\" class=\\"q-scrollarea__thumb q-scrollarea__thumb--v absolute-right q-scrollarea__thumb--invisible\\" style=\\"height: 50px;\\"></div>
</div>"
`;
